stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/querychannels/NonMemberChatEventHandler.kt

```diff
@@ -0,0 +1,139 @@
+package io.getstream.chat.android.offline.querychannels
+
+import io.getstream.chat.android.client.ChatClient
+import io.getstream.chat.android.client.api.models.FilterObject
+import io.getstream.chat.android.client.api.models.QueryChannelsRequest
+import io.getstream.chat.android.client.call.await
+import io.getstream.chat.android.client.events.ChannelUpdatedByUserEvent
+import io.getstream.chat.android.client.events.ChannelUpdatedEvent
+import io.getstream.chat.android.client.events.MemberAddedEvent
+import io.getstream.chat.android.client.events.MemberRemovedEvent
+import io.getstream.chat.android.client.events.NotificationAddedToChannelEvent
+import io.getstream.chat.android.client.events.NotificationMessageNewEvent
+import io.getstream.chat.android.client.events.NotificationRemovedFromChannelEvent
+import io.getstream.chat.android.client.models.Channel
+import io.getstream.chat.android.client.models.Filters
+import io.getstream.chat.android.client.utils.Result
+import io.getstream.chat.android.client.utils.map
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.runBlocking
+
+/**
+ * Default implementation of [ChatEventHandler] which is more generic than [MessagingChatEventHandler]. It skips updates
+ * and makes an API request if a channel wasn't yet handled before when receives [NotificationAddedToChannelEvent],
+ * [NotificationMessageNewEvent], [NotificationRemovedFromChannelEvent].
+ */
+public class NonMemberChatEventHandler(
+    private val client: ChatClient,
+    private val channels: StateFlow<List<Channel>>,
+) : BaseChatEventHandler() {
+
+    /**
+     * Channel filter function. It makes an API query channel request based on cid of a channel and a filter object to
+     * define should be the channel with such cid be in the list of channels or not.
+     */
+    internal val channelFilter: suspend (cid: String, FilterObject) -> Boolean = { cid, filter ->
+        channelsRequest(cid, filter)
+            .map { channels -> channels.any { it.cid == cid } }
+            .let { filteringResult -> filteringResult.isSuccess && filteringResult.data() }
+    }
+
+    private val channelsRequest: suspend (cid: String, FilterObject) -> Result<List<Channel>> = { cid, filter ->
+        client.queryChannelsInternal(
+            QueryChannelsRequest(
+                filter = Filters.and(
+                    filter,
+                    Filters.eq("cid", cid)
+                ),
+                offset = 0,
+                limit = 1,
+                messageLimit = 0,
+                memberLimit = 0,
+            )
+        ).await()
```

Comment:
I see doubling here. What if we could extract it to some object or high level function?

```kotlin
internal object ChannelFilterRequest {
    suspend operator fun invoke(client: ChatClient, cid: String, filter: FilterObject): Result<List<Channel>> =
        client.queryChannelsInternal(
            QueryChannelsRequest(
                filter = Filters.and(
                    filter,
                    Filters.eq("cid", cid)
                ),
                offset = 0,
                limit = 1,
                messageLimit = 0,
                memberLimit = 0,
            )
        ).await()
}

```

Then we can reuse this object in both event handlers